src/lib.rs

```diff
@@ -1,6 +1,159 @@
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
+extern crate memmap;
+extern crate byteorder;
+extern crate time;
+#[macro_use] extern crate bitflags;
+#[cfg(unix)] extern crate nix;
+#[cfg(windows)] extern crate kernel32;
+#[cfg(test)] extern crate rand;
+
+use byteorder::{LittleEndian, WriteBytesExt};
+use memmap::{Mmap, Protection};
+use std::ffi::CString;
+use std::fs::OpenOptions;
+use std::io::Cursor;
+use std::io::prelude::*;
+
+const HDR_LEN: usize = 40;
+const TOC_BLOCK_LEN: usize = 16;
+
+type Endian = LittleEndian;
+
+#[cfg(unix)]
+fn get_process_id() -> i32 {
+    nix::unistd::getpid()
+}
+
+#[cfg(windows)]
+fn get_process_id() -> i32 {
+    unsafe { kernel32::GetCurrentProcessId() as i32 }
+}
+ 
+bitflags! {
+    /// Flags used to modify how a client exports metrics
+    pub struct MMVFlags: u32 {
+        /// Metric names aren't prefixed with MMV filename
+        const NOPREFIX = 1;
+        /// PID check is needed
+        const PROCESS  = 2;
+        /// Allow for "no value available" values
+        const SENTINEL = 4;
+    }
+}
+
+/// Client used to export metrics
+pub struct Client {
+    path: String,
+    flags: MMVFlags,
+    cluster_id: u32,
+}
+
+impl Client {
+    /// Creates a new client with default flags and cluster ID `0`
+    pub fn new(path: &str) -> Client {
+        Client::new_custom(path, MMVFlags::empty(), 0)
+    }
+
+    /// Creates a new client with flags and cluster ID
+    pub fn new_custom(path: &str, flags: MMVFlags, cluster_id: u32) -> Client {
+        Client {
+            path: path.to_owned(),
+            flags: flags,
+            cluster_id: cluster_id,
+        }
     }
+
+    /// Exports metrics by writing to an MMV file
+    pub fn export(&self) -> std::io::Result<()> {
+        let mut file = OpenOptions::new().read(true).write(true).create(true)
+            .open(&self.path)?;
+        let mmv_size = HDR_LEN + 2*TOC_BLOCK_LEN;
+        file.write(&vec![0; mmv_size])?;
+
+        let mut mmap = Mmap::open(&file, Protection::ReadWrite)?;
+        let mut cursor = Cursor::new(unsafe { mmap.as_mut_slice() });
+
+        // MMV\0
+        cursor.write_all(CString::new("MMV")?.to_bytes_with_nul())?;
```

Comment:
please refactor the code from here into separate functions for writing header and toc blocks